src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -891,7 +922,27 @@ public static Delegate GetDelegateForFunctionPointer(IntPtr ptr, Type t)
 
         public static TDelegate GetDelegateForFunctionPointer<TDelegate>(IntPtr ptr)
         {
-            return (TDelegate)(object)GetDelegateForFunctionPointer(ptr, typeof(TDelegate));
+            if (ptr == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(ptr));
+            }
+
+            Type t = typeof(TDelegate);
+            if (t.IsGenericType)
+            {
+                throw new ArgumentException(SR.Argument_NeedNonGenericType, nameof(TDelegate));
+            }
+
+            // COMPAT: This block of code isn't entirely correct.
+            // Users passing in typeof(MulticastDelegate) as 't' skip this check
+            // since Delegate is a base type of MulticastDelegate.
+            Type? c = t.BaseType;
+            if (c != typeof(Delegate) && c != typeof(MulticastDelegate))
+            {
+                throw new ArgumentException(SR.Arg_MustBeDelegate, nameof(TDelegate));
+            }
+
+            return (TDelegate)(object)GetDelegateForFunctionPointerInternal(ptr, t);
         }
 
         [RequiresDynamicCode("Marshalling code for the delegate might not be available. Use the GetFunctionPointerForDelegate<TDelegate> overload instead.")]
```

Comment:
Nit: I just tried this and noticed that there is a double-dot printed in the warning message `Use the GetFunctionPointerForDelegate<TDelegate> overload instead..`. Should we stop appending `.` to the message in ReflecitonMethodBodyScanner.cs?

Reply:
I think we'll never be consistent with this. It might be better to just fix this up when the message is generated (append the dot only if it's not there).